storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+	}
+
+	*(host->base + SDHOST_REG_INTR_STATUS_ENABLE) = 0;
+	*(host->base + SDHOST_REG_INTR_SIGNAL_ENABLE) = 0;
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) = 0;
+}
+
+
+void sdcard_free(unsigned int slot)
+{
+	GET_HOST_OR_RETURN();
+	if (!host->sdioInitialized) {
+		return;
+	}
+
+	mutexLock(host->eventLock);
+	_sdcard_free(host);
+	mutexUnlock(host->eventLock);
+	resourceDestroy(host->eventLock);
+	host->eventLock = 0;
+}
+
+
+int sdcard_initHost(unsigned int slot)
+{
+	if (!presenceEvents.initialized) {
+		if (mutexCreate(&presenceEvents.lock) < 0) {
+			return -ENOMEM;
+		}
+
+		if (condCreate(&presenceEvents.cond) < 0) {
+			resourceDestroy(presenceEvents.lock);
+			return -ENOMEM;
+		}
+
+		presenceEvents.initialized = true;
+	}
+
+	GET_HOST_OR_RETURN(-ENOENT);
+
+	if (host->sdioInitialized) {
+		return (initializedHosts < PLATFORM_SDIO_N_HOSTS) ?
+			(PLATFORM_SDIO_N_HOSTS - initializedHosts) :
+			0;
+	}
+
+	host->cmdLock = 0;
+	host->eventCond = 0;
+	host->eventLock = 0;
+
+	/* Perform platform-specific configuration */
+	sdio_platformInfo_t info;
+	if (sdio_platformConfigure(slot, &info) < 0) {
+		return -EIO;
+	}
+
+	host->refclkFrequency = info.refclkFrequency;
+	host->isCDPinSupported = info.isCDPinSupported;
+	host->isWPPinSupported = info.isWPPinSupported;
+
+	/* Map register bank into our virtual memory */
+	void *ptr = mmap(NULL, _PAGE_SIZE, PROT_WRITE | PROT_READ, MAP_DEVICE, OID_PHYSMEM, info.regBankPhys);
+	if (ptr == MAP_FAILED) {
+		_sdcard_free(host);
+		return -EIO;
+	}
+
+	host->base = ptr;
+	if (sdhost_allocDMA(host) < 0) {
+		_sdcard_free(host);
+		return -ENOMEM;
+	}
+
+	if (mutexCreate(&host->cmdLock) < 0) {
+		_sdcard_free(host);
+		return -ENOMEM;
+	}
+
+	if (sdhost_reset(host, CLOCK_CONTROL_RESET_ALL) < 0) {
+		_sdcard_free(host);
+		return -EIO;
+	}
+
+	if (sdcard_configClockAndPower(host, SD_FREQ_INITIAL) < 0) {
+		_sdcard_free(host);
+		return -EIO;
+	}
+
+	if (!host->isCDPinSupported) {
+		/* Spoof card state so that it always appears inserted */
+		*(host->base + SDHOST_REG_HOST_CONTROL) |= HOST_CONTROL_CARD_DET_TEST;
+		*(host->base + SDHOST_REG_HOST_CONTROL) |= HOST_CONTROL_CARD_DET_TEST_ENABLE;
+	}
+
+	*(host->base + SDHOST_REG_INTR_STATUS_ENABLE) = SDHOST_STATUS_MASK;
+	if (sdcard_startEventISR(host, info.interruptNum) < 0) {
+		_sdcard_free(host);
+		return -ENOMEM;
+	}
+
+	host->sdioInitialized = true;
+	initializedHosts++;
+	return (initializedHosts < PLATFORM_SDIO_N_HOSTS) ?
+		(PLATFORM_SDIO_N_HOSTS - initializedHosts) :
+		0;
+}
+
+
+/* Calculate the divisor value so the output frequency is no larger than `freq` */
+static int sdcard_calculateDivisor(uint32_t refclk, uint32_t freq, uint32_t *divisor)
+{
+	if (freq >= refclk) {
+		*divisor = CLOCK_CONTROL_DIV_1;
+		return 0;
+	}
+	else {
+		uint32_t val;
+		for (val = CLOCK_CONTROL_DIV_2; val <= CLOCK_CONTROL_DIV_256; val <<= 1) {
+			refclk /= 2;
+			if (freq >= refclk) {
+				*divisor = val;
+				return 0;
+			}
+		}
+
+		return -1;
+	}
```

Comment:
`else` is redundant, early return in `if` anyway
```suggestion
	uint32_t val;
	for (val = CLOCK_CONTROL_DIV_2; val <= CLOCK_CONTROL_DIV_256; val <<= 1) {
		refclk /= 2;
		if (freq >= refclk) {
			*divisor = val;
			return 0;
		}
	}

	return -1;
```